### PR TITLE
[MAINT] Clean CI tools

### LINKED ISSUE #2746: restore flake8 and remove unused ci scripts; merge tools and continuous_integration

### PR DESCRIPTION
**Pull request by NicolasGensollen**
Closes #2746 

This PR proposes to do some cleaning of the CI utilities:

- Reorganize CI scripts in folder `build_tools` rather than having things in both `tools` and `continuous_integration`. If a script is dedicated to a specific platform, it is then placed in the corresponding sub-folder. For example, all scripts for CircleCI are in `build_tools/circle/`, for Github actions `build_tools/github/`...
- Rename files to make their name shorter and/or clarify their purpose
- Add some documentation for configuring how CircleCI will build the doc through commit messages (we can include that later in the contributing docs once Thomas's PR gets merged)
- Remove unused scripts and files (previously used for caching on CircleCI)
- Update flake8_diff script and make it run through GitHub Actions (currently runs with the min dependency job, but we can change that I needed)
- Run `make test-doc` which was not executed on the CI since we dropped testing on Travis.
- Create a `maint_tools` folder to hold other maintenance scripts

### CODE REVIEW

**Review by bthirion — APPROVED** [review]
LGTM, but I'm not sure how to check all details ...

File: .circleci/README.md [hunk 1]
```diff
@@ -1,14 +1,38 @@
...
+
+You can request a CircleCI full build from a Pull Request at any time by including the tag "[circle full]" in your commit message. Note that this will trigger a full build of the documentation which usually takes around 90 minutes.
+
+```bash
+$ git commit -m "[circle full] request full build"
+```
+
+## Dataset caching
+
+We also implemented a dataset caching strategy within the CircleCI workflow such that datasets are only downloaded once every week. Once these datasets are cached, they will be used by all jobs running on CircleCI without requiring any download. This saves a lot of time and avoid potential network errors which can happen when downloading datasets from remote servers.
```
Comment (bthirion) [inline .circleci/README.md:21]:
```suggestion
We also implemented a dataset caching strategy within the CircleCI workflow such that datasets are only downloaded once every week. Once these datasets are cached, they will be used by all jobs running on CircleCI without requiring any download. This saves a lot of time and avoids potential network errors that can happen when downloading datasets from remote servers. 
```

**Review by bthirion — COMMENTED** [review]
minor details.

File: .circleci/README.md (hunk 1 above)
Comment (NicolasGensollen) [reply .circleci/README.md]: Thanks!

**Comment by jeromedockes** [comment]
> LGTM, but I'm not sure how to check all details ...

same; I suggest that when @NicolasGensollen considers this ready (maybe now?) we merge it and we can open a couple of dummy PRs if we want to check something works as expected, eg the flake8 on the diff

**Comment by NicolasGensollen** [comment]
>I suggest that when @NicolasGensollen considers this ready (maybe now?) we merge it and we can open a couple of dummy PRs if we want to check something works as expected, eg the flake8 on the diff

@jeromedockes I consider this PR as ready indeed. So, unless you see something wrong, I think we can merge it and see if everything works as expected.

**Comment by jeromedockes** [comment]
great! merging then

**Pull request merged**